tools/schemacode/bidsschematools/validator.py

```diff
@@ -714,9 +780,9 @@ def _get_directory_suffixes(my_schema):
 
 def validate_bids(
     bids_paths,
+    dummy_paths=False,
```

Comment:
I don't like this name -- there is nothing "dummy" about them...  I would like to make separation of two modes to be made more "explicit", not a flag in API.
I was to propose a change, but let me first clarify - do you foresee to validation of a collection of individual folders, e.g. `['sub-01', 'sub-02']` -- would that be legit? 
I kinda like an idea to be able to validate a specific subject[/session] but I wonder if that would be feasible in the long run as more checks would be added -- would it mean to not check anything about e.g. `rawdata/` or `participants.tsv` at the top level?

Reply:
> > there is nothing "dummy" about them
> 
> I think the name is perfectly accurate: [“dummy, noun. An imitation of a real or original object, intended to be used as a practical substitute.”](https://www.wordnik.com/words/dummy)

it doesn't matter if they are *real* or *substitute* since this particular validator does not care to look into content of the files. They can as well/better be *real* paths! The critical difference IMHO that they are paths to files, and thus are not recursively navigated to pick up more paths via `_get_paths` (note that name `_get_paths` is not descriptive of what it means to get_paths from the paths it gets... may be `_get_files_recursively` would have been a bit more descriptive).  

So, in principle, if all `paths` point to files not directories, there should be no point in calling `_get_paths`, right?  So then one solution could be to do `isdir` on each path and if a directory, and only if dir, expand into a list of files under it.  And then this flag could be avoided entirely.

But I don't like this suggestion of mine because it makes it somewhat "implicit" and thus in some cases would consider some `sub-01` user-provided folder to be a filename depending on the status of filesystem/invocation.  That is why I would have preferred a more explicit separation:

There could be `_check_filenames` and `validate_bids` (already there, entry point function for a user) where 
- `_check_filanemes` - would expect to operate only on filenames relative to the top of the dataset (e.g. `sub-01/func/sub-01_bold.nii.gz'), and not require any content/physical presence. Consider it the test for `NOT_INCLUDED` (1) of current bids-validator errors.  I used `_check` to initiate collection of `_checks` which validator should slowly grow. E.g. you could right away add a `_check_valid_json` which would check for `JSON_INVALID` error condition.
- `validate_bids` would call `_get_paths` to expand/filter, figure out the top of the dataset, convert to relative paths and then call `_check_filenames` or whatever other checks it would know about.

Reply:
> it doesn't matter if they are real or substitute since this particular validator does not care to look into content of the files.

it matters, because while the validator does not check the files (yet) it needs to check whether they are files or directories.

> They can as well/better be real paths!

yes, the can, and ideally should. This entire thing was introduced because you wanted us to test non-existent paths for validator inheritance. It is however also useful for the DANDI API evaluation, as well as other similar use cases people might want to use the library for, so it seems worthwhile in addition to the original rationale.

> So, in principle, if all `paths` point to files not directories, there should be no point in calling `_get_paths`, right?

Yes, except when they are dummy paths, i.e. not real paths, in that case you cannot check either isfile or isdir

> in some cases would consider some sub-01 user-provided folder to be a filename

yes, that is the issue with just checking for isdir and assuming that everything which is not a dir, including a dummy path, is actually a file.

I'm not entirely sure why we should separate this into a separate top-level function, in effect what we are doing is simply ignoring `NOT_INCLUDED` errors... perhaps the better plan would be to parameterize ignoreable errors once they are codified (though that could get convoluted). In any case, for the time being, since we operate on a list of paths in the end anyway, that simply means skipping the “smart” path acquisition bits, which is what we now do.

Reply:
> it matters, because while the validator does not check the files (yet) it needs to check whether they are files or directories.

so - "not yet", and is that check a part of regex? if not -- could be a separate `_check_pathtypes` check AFAIK. If part of regex (directories must have trailing `/`) -- then again no need to incorporate `isdir` within `_check_filenames` (might better become `_check_paths` then) - assignment of trailing `/` would be done by `_get_paths` based on the "type" (dir vs file)

> ... , in that case you cannot check either isfile or isdir

again -- you are not doing that discrimination "yet" or at all in the regex based checking.  I think it is valuable to keep design/implementation clean as if you have checks which are based purely on regexes - keep it that way, do not mix in all kinds of other checks.  If you had some other implementation, we would consider different design.  ATM I am just asking about cleaner separation of regex matching functionality which does not need any real path/directory.

> I'm not entirely sure why we should separate this into a separate top-level function

because then you can unittest that function much better by itself, instead of testing the entire larger beast, and require instantiation of real paths etc.  Moreover later we would be able to annotate checks either they need real paths/directories or not, thus making it easy to go through checks which do not require real paths if we do not have them.

Note -- for DANDI archive we would indeed need entry into validation which would provide just the paths and not just to checking by regexes. So we indeed might need another function or options to `validate_bids` to operate  solely on a list of paths, and then use only checks which do not need real files.  But we better arrive to that point, and again I would not call those paths "dummy" -- they would be real paths for real files but "on the archive" i.e. just not available locally.

Reply:
> because then you can unittest that function much better by itself, instead of testing the entire larger beast, and require instantiation of real paths etc.

But passing a parameter which controls whether directory walk is invoked prevents us from testing those parts of the “larger beast”. Speaking of larger beasts, there are nevertheless parts of the function, such as bids schema dir loading and reporting which would need to remain not equal (repeated code), but identical (splitting off a function for the commonalities between the two top-level functions). That would make everything more nested overall, and I think would contravene rather than support “cleaner separation”.

> keep design/implementation clean as if you have checks which are based purely on regexes

All checks are always based purely on regexes, there is no exception and I don't think there should ever to be one, even when we end up checking file contents — that would be a subsequent check after we get the regex match paths.

> they would be real paths for real files but "on the archive"

That is specific to the DANDI use case, as far as the validator is concerned those paths are just strings with no meaningful parent directories and therefore no known location in reality :) For the validator, the main utility of such an option would be automatically generated or hard-coded strings for testing (as are already present in this PR), in which case they are purely theoretical dummy paths.

In any case I think we have long departed the realm of inheritance support, and I would need some of these features to fix active bugs (e.g. https://github.com/bids-standard/bids-specification/issues/1155 ). If you insist on the refactoring, could you perhaps open an issue about that? We could also use this opportunity to get broader feedback.